torch_np/_detail/_reductions.py

```diff
@@ -0,0 +1,165 @@
+""" Implementation of reduction operations, to be wrapped into arrays, dtypes etc
+in the 'public' layer.
+
+Anything here only deals with torch objects, e.g. "dtype" is a torch.dtype instance etc
+"""
+
+import torch
+from . import _util
+from . import _scalar_types
+
+NoValue = None
+
+
+def _atleast_float(dtype, other_dtype):
+    """Cast bools and ints and floats to a default float; complex to default complex.
+    """
+    if dtype is None:
+        dtype = other_dtype
+    if not dtype.is_floating_point:
+        if dtype.is_complex:
+            sctype = _scalar_types.default_complex_type
+        else:
+            sctype = _scalar_types.default_float_type
+        dtype = sctype.torch_dtype
+    return dtype
+
+
+def count_nonzero(a, axis=None):
+    # XXX: this all should probably be generalized to a sum(a != 0, dtype=bool)
+    try:
+        return a.count_nonzero(axis)
+    except RuntimeError:
+        raise ValueError
+    return tensor
+
+
+def argmax(tensor, axis=None, out=None, *, keepdims=NoValue):
+    axis = _util.allow_only_single_axis(axis)
+    tensor = torch.argmax(tensor, axis)
+    return tensor
+
+def argmin(tensor, axis=None, out=None, *, keepdims=NoValue):
+    axis = _util.allow_only_single_axis(axis)
+    tensor = torch.argmin(tensor, axis)
+    return tensor
+
+
+def any(tensor, axis=None, *, where=NoValue):
+    if where is not None:
+        raise NotImplementedError
+
+    axis = _util.allow_only_single_axis(axis)
+
+    if axis is None:
+        result = tensor.any()
+    else:
+        result = tensor.any(axis)
+    return result
+
+
+def all(tensor, axis=None, *, where=NoValue):
+    if where is not None:
+        raise NotImplementedError
+
+    axis = _util.allow_only_single_axis(axis)
+
+    if axis is None:
+        result = tensor.all()
+    else:
+        result = tensor.all(axis)
+    return result
+
+
+def max(tensor, axis=None, initial=NoValue,
+         where=NoValue):
+    if where is not None:
+        raise NotImplementedError
+    if initial is not None:
+        raise NotImplementedError
```

Comment:
minor: the form
```python
    if initial is not None or where is not None:
        raise NotImplementedError
```
as used further down seems nicer

Reply:
same for `min` a few liner lower

Reply:
Thanks, done